### ISSUE TITLE
Xamarin.iOS app freezes while navigating

### ISSUE THREAD

**Issue by bertt** [Mapsui/Mapsui#1191]
I've created a Xamarin.Forms app containing Mapsui map with 1 graphics layer. When testing the iOS version on IPhone a complete app freeze can happen while navigating the map. 

I've made a project to reproduce, see project 'MemoryProvider' (branch bug/getfeaturesinview_not_called/) in the [samples/Mapsui.Samples.MemoryProvider folder](https://github.com/Mapsui/Mapsui/tree/bug/getfeaturesinview_not_called/Samples/Mapsui.Samples.MemoryProvider/MemoryProvider/). Project under test is MemoryProvider.iOS. Note, in this branch HasFling is set to false.

Its not easy to reproduce, its just a matter of continuous pan and zoom action's while graphic layer is visible (layer has MaxVisible defined), normally within 50 map actions the freeze happens.

Some more observations:

- just a map without the graphics layer: no freeze happens (so far);
- settings like map.RotationLock  and map.MyLocationEnabled don't influence the app freeze.

For completeness, an screenshot after freeze. Note the graphics on the right/bottom side stopped drawing.

![image](https://user-images.githubusercontent.com/538812/132106343-06262196-eb90-4499-b7fb-bbcd81c9abfb.png)

**Comment by charlenni** [comment]
Could you try, what's happening if you use MapControl instead of MapView? Perhaps you have to remove some special things (like MyLocationLayer) but it would be interesting, if this chances something.

**Comment by bertt** [comment]
Good suggestion, it still happens when changing mapsui:MapView to mapsui:MapControl in MainPage.xaml. But the freeze seems to occur less frequently.

**Comment by bertt** [comment]
Crash log from phone:

```
Date/Time:           2021-09-05 09:52:52.5998 +0200
Launch Time:         2021-09-05 09:52:03.1987 +0200
OS Version:          iPhone OS 14.7.1 (18G82)
Release Type:        User
Baseband Version:    4.04.00
Report Version:      104

Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Reason: Namespace SPRINGBOARD, Code 0x8badf00d
Termination Description: SPRINGBOARD, <RBSTerminateContext| domain:10 code:0x8BADF00D explanation:scene-update watchdog transgression: application<com.mapsui.MemoryProvider>:59325 exhausted real (wall clock) time allowance of 10.00 seconds | ProcessVisibility: Background | ProcessState: Running | WatchdogEvent: scene-update | WatchdogVisibility: Background | WatchdogCPUStatistics: ( | "Elapsed total CPU time (seconds): 4.980 (user 4.980, system 0.000), 8% CPU", | "Elapsed application CPU time (seconds): 0.386, 1% CPU" | ) reportType:CrashLog maxTerminationResistance:Interactive>
Triggered by Thread:  0
```

**Comment by bertt** [comment]
From console logs kernel fault messages:

```
IOReturn IOGPUDevice::new_resource(IOGPUNewResourceArgs *, struct IOGPUNewResourceReturnData *, IOByteCount, uint32_t *): PID 59429 likely leaking IOGPUResource (count=56000)
IOReturn IOGPUDevice::new_resource(IOGPUNewResourceArgs *, struct IOGPUNewResourceReturnData *, IOByteCount, uint32_t *): PID 59429 likely leaking IOGPUResource (count=55000)
```

**Comment by charlenni** [comment]
If you set UseGPU=false for MapControl before it is created and test it again, the crash happens too?

**Comment by charlenni** [comment]
Do we have a memory problem? You create at each drawing process hundreds of Polygons. They are never reused. Perhaps a dictionary would be helpful to store the already created Polygons? I don't know, if SkiaSharp interface stores some pointers to such structures.

**Comment by charlenni** [comment]
In moment we refresh the MapControl each time we get a new position for the map, because this generates a refresh. Now think about what could happen when you move around the map. We get in the range of sub ms new refresh's. Each of them generates a new bunch of Polygons, only to through them away a few microseconds later.

**Comment by pauldendulk** [comment]
> In moment we refresh the MapControl each time we get a new position for the map, because this generates a refresh. Now think about what could happen when you move around the map. We get in the range of sub ms new refresh's. Each of them generates a new bunch of Polygons, only to through them away a few microseconds later.

This is inefficient wrt performance but if there is no memory leak this should not crash the app.

**Comment by pauldendulk** [comment]
Perhaps we add some more logging of thing related to memory use. Perhaps:

```Logger.Log(LogLevel.Debug, $"Memory after rasterizing layer {GC.GetTotalMemory(true):N0}");```

And there may be a way to request memory or feature use in SkiaSharp.

**Comment by bertt** [comment]
> If you set UseGPU=false for MapControl before it is created and test it again, the crash happens too?

With UseGPU = false the crash does not occur. So it seems to be related to drawing a lot of graphics on SKGLView (GPU) and not with SKCanvasView (CPU). This SkiaSharp issue seems quite related https://github.com/mono/SkiaSharp/issues/1683

**Comment by charlenni** [comment]
I tried to get some more informations, but didn't find anything except the issue of SkiaSharp you mentioned.

I assume, that this is a problem with the deprecated OpenGL on iOS. Perhaps we have to wait until SkiaSharp uses Metal as backend (perhaps in 2.88.x?). Until than, setting `UseGPU=false` on iOS should do the trick. But I don't have an idea about the performance loose.

**Comment by bertt** [comment]
Maybe we can force UseGPU=false when building for iOS (so people don't have to know about the UseGPU flag when building for iOS).

#if __IOS__
UseGPU=false
#endif

**Comment by charlenni** [comment]
BTW did you test the same with the Mac version of Mapsui.Forms?

**Comment by bertt** [comment]
> BTW did you test the same with the Mac version of Mapsui.Forms?

No, so far only on IPhone

**Comment by charlenni** [comment]
I don't know, how much this reduces the performance and reaction on iOS. The smartphone's runtime could also be affected.

So it would be better, if this isn't a default change.

**Comment by pauldendulk** [comment]
[This](https://github.com/mono/SkiaSharp/issues/1683#issuecomment-914139405) reply indicates that just drawing will eventually crash the app. I'd say we should always worse performance over crash. Also it will most users will be unaware of it and find out late after te release of their app. So often it will not be their conscious choice. 

@Sebastian1989101 Are you using GPU on iOS? Encountered such a problem?

**Comment by pauldendulk** [comment]
A [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) might helpt to get this resolved.

**Comment by charlenni** [comment]
Checks had to be:
1. Do you get it with MapControl too?
2. Do you get it with the native iOS MapControl too?
3. Could you reproduce the problem with a SKGLView without Mapsui?

**Comment by charlenni** [comment]
Why?
1. If it is a problem with MapView, then it could be a problem if removing/adding the layers
2. If it is a problem with the MapControl of Forms, then it could be a problem with the Forms components of SkiaSharp
3. The same as 2.

**Comment by charlenni** [comment]
Why?
1. If it is a problem with MapView, then it could be a problem if removing/adding the layers
2. If it is a problem with the MapControl of Forms, then it could be a problem with the Forms components of SkiaSharp
3. The same as 2.

**Comment by Sebastian1989101** [comment]
> @Sebastian1989101 Are you using GPU on iOS? Encountered such a problem?

@pauldendulk if `UseGPU = true` was always the default in Mapsui 2.0, then it was also set this way in my apps on iOS. But now, I did not encounter this bug yet even with ~13000 data points on the map seperated in ~150 layers. I only had lags when changing the layer collection but nothing else and I tested my apps on multiple devices before each release (iPhone 4S on iOS 9.0 as the oldest). However I just checked the logs for that specific app and there are 67 crashes with a pretty similar crash log to this one with ~52000 daily active users average over the last 30 days, that does not seems to be that impactful. But maybe I also trigger to few redraws from changes.

**Comment by pauldendulk** [comment]
@Sebastian1989101 Okay thanks. That means it is not as serious as I feared. It could also depend on the kind of feature that is drawn. Bert's sample uses geometries, perhaps you use mainly points.

**Comment by Sebastian1989101** [comment]
@pauldendulk yes, most of my thinks are points. Only a few LineString's are used. However I can recreate the issue from the linked SkiaSharp issue with the memory. If I set `UseGPU = false` the app uses less memory and more consistant compared to `UseGPU = true`. I can't say much about the performance for the latest release tho because I still use my local modified version currently (because without the mass edits of the LayerCollection the loading is just too long on changes in this app).

**Comment by pauldendulk** [comment]
If anyone feels like it we could create a minimal reproducible sample and post it in skiasharp.

**Comment by pauldendulk** [comment]
GPU does not work on all platforms. We tried to set sensible defaults (selecting individual platforms). If not the user can override it with MapControl.UseGPU = false;
